### PR TITLE
Use username during nb spawning

### PR DESCRIPTION
**Pull request by maroroman**
Resolves: #564

## Description
This PR just adds the username to the enable notebook request to utilise features added in https://github.com/opendatahub-io/odh-dashboard/pull/645 and related PRs

## How Has This Been Tested?
1. Make sure your (admin) PV is present
2. Spawn notebook as impersonated user
3. Verify user's PV is claimed correctly
4. Notebook spawns successfully for targeted user

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by lucferbux** [comment]
Ok, a couple of things, not all related to the changes you made:

- Why the request for all users in `/api/status/:namespace/allowedUsers` is a `POST` rather than a `GET`, even the function is called `getAllowedUsers`, seems weird to me.
- I've found an edge case in `getAllowedUsers`. If there's no admin group or allowed groups selected (this is not ideal, but could be that a group wasn't selected on-prem in the beginning), `getAdminUserList` and `getAllowedUserList` will fail since there will be an empty array `['']` and will return an array of undefined `[ undefined ]`, you can solve this adding another filter to remove empty strings.
- The other thing is that in `getAllowedUserList` we are ignoring the `system:authenticated` group, so if there are multiple users that have spawned notebooks but are not in an specific rhods group, they won't appear in the panel.

CC. @andrewballantyne I would argue that at least the 3rd change is fundamental for the feature, the rest are up for implementation, not sure your thoughts on this.

**Comment by lucferbux** [comment]
> This is by design. I do not know a way we can select those users -- and I'm not entirely sure we want to. There is definitely room to improve here, so I'm open to suggestions.
>
> Effectively this page is to list anyone it knows exists as a user -- if they use "everyone on cluster" I'm a bit weary of trying to crunch that down into potentially 1000s of users.

Yeah, that's what i imagined, getting all the users of the cluster is super overkill. One thing that I was thinking is at least display all the users that have spawned notebooks, as you can get the name in the notebook's labels, but then again, it might add too much logic for that feature.

**Comment by andrewballantyne** [comment]
> One thing that I was thinking is at least display all the users that have spawned notebooks

We do @lucferbux ([code link](https://github.com/opendatahub-io/odh-dashboard/blob/main/backend/src/routes/api/status/adminAllowedUsers.ts#L76))

```
  const returnUsers: AllowedUserMap = {
    ...usersWithNotebooksMap, // users who have notebooks created
    ...allowedUsersMap, // users who are in the allowed list
    ...adminUsersMap, // users who are in the admin list
  };
```

File: backend/src/utils/adminUtils.ts [hunk 1]
```diff
@@ -25,17 +25,17 @@ export const getAdminUserList = async (fastify: KubeFastifyInstance): Promise<st
...
 };
 
 export const getAllowedUserList = async (fastify: KubeFastifyInstance): Promise<string[]> => {
   const allowedGroups = getAllowedGroups();
   const allowedGroupList = allowedGroups
     .split(',')
-    .filter((groupName) => !groupName.startsWith('system:')); // do not include k8s defaults
-
+    .filter((groupName) => !groupName.startsWith('system:')) // do not include k8s defaults
+    .filter((groupName) => groupName != ''); // To prevent edge cases where groups are unset
```
Comment (DaoDaoNoCode) [inline backend/src/utils/adminUtils.ts:38]: Can you combine these two filters together?
`.filter((groupName) => groupName && !groupName.startsWith('system:'))`

File: frontend/src/pages/notebookController/screens/server/StopServerModal.tsx [hunk 2]
```diff
@@ -6,11 +6,16 @@ import useNotification from '../../../../utilities/useNotification';
 import { allSettledPromises } from '../../../../utilities/allSettledPromises';
 
 type StopServerModalProps = {
+  impersonatedUsername: string | undefined;
```
Comment (DaoDaoNoCode) [inline frontend/src/pages/notebookController/screens/server/StopServerModal.tsx:9]:
```suggestion
  impersonatedUsername?: string;
```

File: frontend/src/pages/notebookController/screens/server/StopServerModal.tsx [hunk 3]
```diff
@@ -29,7 +34,10 @@ const StopServerModal: React.FC<StopServerModalProps> = ({ notebooksToStop, onNo
       notebooksToStop.map((notebook) => {
         const notebookName = notebook.metadata.name || '';
         if (!notebookName) return Promise.resolve();
-        return stopNotebook();
+        if (!impersonatedUsername) {
+          impersonatedUsername = notebook.metadata.labels['opendatahub.io/user'];
```
Comment (DaoDaoNoCode) [inline frontend/src/pages/notebookController/screens/server/StopServerModal.tsx:38]: I am not sure whether it's a good practice to manipulating the prop directly, maybe do something like this?
```
remove -> if (!impersonatedUsername) {...}
return stopNotebook(impersonatedUsername ?? notebook.metadata.labels['opendatahub.io/user'])
```

File: frontend/src/pages/notebookController/screens/server/SpawnerPage.tsx [hunk 4]
```diff
@@ -257,7 +258,7 @@ const SpawnerPage: React.FC = () => {
         setSubmitError(e);
         setCreateInProgress(false);
         // We had issues spawning the notebook -- try to stop it
-        stopNotebook().catch(() =>
+        stopNotebook(impersonatedUsername != null ? impersonatedUsername : undefined).catch(() =>
```
Comment (DaoDaoNoCode) [inline frontend/src/pages/notebookController/screens/server/SpawnerPage.tsx:261]:
```suggestion
        stopNotebook(impersonatedUsername ? impersonatedUsername : undefined).catch(() =>
```

File: frontend/src/pages/notebookController/screens/server/SpawnerPage.tsx [hunk 5]
```diff
@@ -364,7 +365,7 @@ const SpawnerPage: React.FC = () => {
           onClose={() => {
             if (currentUserNotebook) {
               const notebookName = currentUserNotebook.metadata.name;
-              stopNotebook()
+              stopNotebook(impersonatedUsername != null ? impersonatedUsername : undefined)
```
Comment (DaoDaoNoCode) [inline frontend/src/pages/notebookController/screens/server/SpawnerPage.tsx:368]:
```suggestion
              stopNotebook(impersonatedUsername ? impersonatedUsername : undefined)
```

File: frontend/src/pages/notebookController/screens/admin/NotebookAdminControl.tsx [hunk 6]
```diff
@@ -115,7 +115,11 @@ const NotebookAdminControl: React.FC = () => {
             </StackItem>
           </Stack>
         </div>
-        <StopServerModal notebooksToStop={notebooksToStop} onNotebooksStop={onNotebooksStop} />
+        <StopServerModal
+          impersonatedUsername={undefined}
```
Comment (DaoDaoNoCode) [inline frontend/src/pages/notebookController/screens/admin/NotebookAdminControl.tsx:119]: If it's an `undefined` you don't even need to pass it.
```suggestion
```

**Review by DaoDaoNoCode — COMMENTED** [review]
left some small nits.

File: frontend/src/pages/notebookController/screens/server/StopServerModal.tsx (hunk 3 above)
Comment (maroroman) [reply frontend/src/pages/notebookController/screens/server/StopServerModal.tsx]: If the same solution still applies then this conversation can be resolved.

**Review by DaoDaoNoCode — APPROVED** [review]
lgtm

**Review by andrewballantyne — APPROVED** [review]
Works well, thanks for bringing back the tab -- it's been a couple releases.

**Pull request merged**